### ISSUE TITLE
Unexpected assignable type on Partialed type

### ISSUE THREAD

**Issue by genesos** [microsoft/TypeScript#41932]
**TypeScript Version:**  4.1.2

**Search Terms:** Unexpected assignable type on Partial

**Expected behavior:** Expect throwing a error. But casting as `A` which has common property `cc`, makes as well.

**Actual behavior:** Error on all assign.

**Related Issues:** Trying to found similar issue, but failed.

**Code**
```ts
interface A {
    a: number;
    cc?: number;
}

interface B {
    b: number;
    cc: number;
}

const a: A = {a: 1};
const b: Partial<B> = a;
//                    ^ "unexpect success"
const c: Partial<B> = {a: 1} as A;
//                    ^ "unexpect success"
const d: Partial<B> = {a: 1};
//                    ^ "error"
```
```Error:(683, 30) TS2322: Type '{ a: number; cc: number; }' is not assignable to type 'Partial<B>'.
  Object literal may only specify known properties, and 'a' does not exist in type 'Partial<B>'.
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
const a = { a: 1 };
const b = a;
//                    ^ "expect error, but success"
const c = { a: 1 };
//                    ^ "expect error, but success"
const d = { a: 1 };
//                    ^ "expected error"

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgILIN4FgBQz-JwBcyIArgLYBG0A3LgcgggPwnnV24C+uuokWIhQAhTAwJV2lGlHp4CzaZzk8+OBAHsQAZzCES6ALyZiyAIzd5W3fqnIACnChhgcADYAeEQD5kJuHkAeiDGMPCwgD1kACIIAA8ABwgEfWgoTSgAGmQqMn0dMmYIHR0Y3Bs9JhInFzcvX39TEktCHTRg0Ijuxmi4pJS0qAzs3PzkQuLS8o1tKoATGudXD28-EwwzS06e3eQ+hOTUiHnkdMyYoA)

**Comment by MartinJohns** [comment]
This works as expected.

Assignment to `b` works, because TypeScript is structurally typed, and the structure of `a` is compatible. Everything that must be present according to `Partial<B>` is present in `A`, and nothing in `A` conflicts with the type `Partial<B>`.

Assignment `c` is essentially the same as assignment `b`. You tell the compiler to treat your object literal as type `A`, so you don't get the error as you do in assignment `d`. Keep in mind that type assertions, often called casts, are usually a way to tell the compiler to ignore type errors. IMO you should avoid using them.

**Comment by genesos** [comment]
@MartinJohns 
Thanks for your comment! but how can I block assignment of object which has unused property?
For example,

```ts
interface A {
    a: number;
    cc?: number;
}
interface B {
    b: number;
    cc: number;
}
type NoPropertyA = Omit<Partial<B>, 'a'>; // How???
const noPropertyA:NoPropertyA = {a: 1} as A;
//                              ^ Success, but error needed.
```

**Comment by MartinJohns** [comment]
> but how can I block assignment of object which has unused property?

As TypeScript is a structurally typed language... you can't, generally. You can't prevent the presence of additional properties. You would want #12936 for this, but don't expect it to be implemented any time soon.

In your last example... Just drop the `as A`, it's not needed and will just hide bugs.
